Source/JavaScriptCore/runtime/JSStringInlines.h

```diff
@@ -72,6 +72,88 @@ ALWAYS_INLINE bool JSString::equalInline(JSGlobalObject* globalObject, JSString*
     return WTF::equal(str1, str2, length);
 }
 
+JSString* JSString::tryReplaceOneCharImpl(JSGlobalObject* globalObject, UChar search, JSString* replacement, uint8_t* stackLimit, bool& found)
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    if (UNLIKELY(bitwise_cast<uint8_t*>(currentStackPointer()) < stackLimit))
+        return nullptr; // Stack overflow
+
+    if (this->isNonSubstringRope()) {
+        JSRopeString* rope = static_cast<JSRopeString*>(this);
+        JSString* oldFiber0 = rope->fiber0();
+        JSString* oldFiber1 = rope->fiber1();
+        JSString* oldFiber2 = rope->fiber2();
+
+        ASSERT(oldFiber0);
+        JSString* newFiber0 = oldFiber0->tryReplaceOneCharImpl(globalObject, search, replacement, stackLimit, found);
+        RETURN_IF_EXCEPTION(scope, nullptr);
+        if (UNLIKELY(!newFiber0))
+            return nullptr;
+        if (found)
+            RELEASE_AND_RETURN(scope, jsString(globalObject, newFiber0, oldFiber1, oldFiber2));
+
+        if (oldFiber1) {
+            JSString* newFiber1 = oldFiber1->tryReplaceOneCharImpl(globalObject, search, replacement, stackLimit, found);
+            RETURN_IF_EXCEPTION(scope, nullptr);
+            if (UNLIKELY(!newFiber1))
+                return nullptr;
+            if (found)
+                RELEASE_AND_RETURN(scope, jsString(globalObject, oldFiber0, newFiber1, oldFiber2));
+        }
+
+        if (oldFiber2) {
+            JSString* newFiber2 = oldFiber2->tryReplaceOneCharImpl(globalObject, search, replacement, stackLimit, found);
+            RETURN_IF_EXCEPTION(scope, nullptr);
+            if (UNLIKELY(!newFiber2))
+                return nullptr;
+            if (found)
+                RELEASE_AND_RETURN(scope, jsString(globalObject, oldFiber0, oldFiber1, newFiber2));
+        }
+
+        return this; // Not found.
+    }
+
+    auto thisView = this->view(globalObject);
+    RETURN_IF_EXCEPTION(scope, nullptr);
+
+    size_t index = thisView->find(search);
+    if (index == WTF::notFound)
+        return this; // Not found.
+    found = true;
+
+    // Case 1: The matched character is the only character in the string.
+    unsigned length = thisView->length();
+    if (length == 1)
+        return replacement;
+
+    // Case 2: The matched character is the last character in the string.
+    JSString* left = jsSubstring(globalObject, this, 0, index);
```

Comment:
We should avoid creating `left` when it may not be used. `index == 0` case does not use it.